Retire Stake and Minimum stake

ii-cruz opened the pull request:
## What's in this pull request?

- Retire stake was added to the staker. It is required so that we are sure there are enough funds to deduct the remove stake fee on a commit failed transaction.
- Introduces the Minimum Stake of 100 NIM. This is to prevent a blown-up state size for the Staking Contract. Thus we add two minimum stake invariants to the staker: 
Total balance of a staker (active + inactive + retire balances) >= Minimum Stake OR == 0
Non retired balances of a staker (active + inactive balances) >= Minimum Stake OR == 0
- Adds all the test case scenarios testing the new invariants and the new retired balance state.
- Restricts the remove stake to only allow the removal of all retired balance.
- Finishes renaming all `unstake` transactions to `remove_stake` to make it standard across the codebase.
- Finishes renaming all `stake` transactions to `add_stake` to make it standard across the codebase.

#### This fixes #2000.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

ii-cruz committed:
Min stake

Introducing the minimum active stake in addition to the minimum total stake.
Added more intrinsic checks on the staking contract incoming txs.

ii-cruz committed:
Retire stake

Retire stake tests
Fixing reserve_balance for staker and validator and adding tests.

sisou commented on web-client/src/transaction.rs:
```diff
@@ -837,6 +846,14 @@ pub struct PlainSetActiveStakeData {
     pub new_active_balance: u64,
 }
 
+/// JSON-compatible and human-readable format of set inactive stake data.
```
Copy-pasta
```suggestion
/// JSON-compatible and human-readable format of retire stake data.
```

sisou replied:
Please also fix the comment for `PlainSetActiveStakeData`, to read "set active stake data".

ii-cruz replied:
Fixed! thank you so much for the good eye🕵️‍♂️

sisou commented:
From only reading the PR description, not any code yet: these to conditions seem like they could be simplified:

> Total balance of a staker (active + inactive + retire balances) >= Minimum Stake OR == 0
> Non retired balances of a staker (active + inactive balances) >= Minimum Stake OR == 0

Since the second condition is a subset of the first, the first condition can be reduced to simply "retired balance >= minimum stake OR == 0" and does not have to include the two other ones.

sisou commented on primitives/src/policy.rs:
```diff
@@ -105,6 +105,12 @@ impl Policy {
+    /// This minimum is applied to the amount of:
+    ///     - non retired stake
+    ///     - total stake balance
+    pub const MINIMUM_STAKE: u64 = 10_000_000;
```
Offtopic comment:
I always find this way of separating by thousands weird, because 10k luna doesn't tell me anything useful. I would prefer this whole file separates by 5 zeros at the end first, and then the NIM part by thousands:
```suggestion
    pub const MINIMUM_STAKE: u64 = 100_00000;
```

hrxi replied:
I guess it'd need a separate PR, but I generally agree.

sisou commented on docs/becoming_validator.md:
```diff
@@ -184,12 +184,12 @@ validator:inactivate <validator_address> <validator_address> <signing_private_ke
 
 
-### 5.2 Unstake delegated Stake
+### 5.2 Remove Stake
```
I don't quite understand this step in the documentation. Is it for validators, or for stakers? In any case, the `retire` step is missing.

ii-cruz replied:
conclusion: this needs updating but later. Currently, the documentation seems to be confusing the validator and stake inactivation. However, we need to first update the arpl and then update this file.

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -9,6 +9,7 @@ use nimiq_primitives::coin::Coin;
 use nimiq_primitives::policy::Policy;
 use serde::{Deserialize, Serialize};
 
+use crate::RetireStakeReceipt;
```
Can this import be moved into the `use crate` group below?

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -21,11 +22,13 @@ use crate::{
-/// 2. Stake:            Adds coins from any outside address to a staker's active balance.
+/// 2. AddStake:         Adds coins from any outside address to a staker's active balance.
+///                      This action is only possible if:
+///                        (a) if the resulting non retired funds respect the invariant 1 - minimum stake for non retired funds.
```
Double "if" (already present in the intro above).
```suggestion
///                        (a) the resulting non-retired funds respect the invariant 1 - minimum stake for non-retired funds.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -35,17 +38,34 @@ use crate::{
-///                      Only inactive stake that has been released can be withdrawn (*).
+/// 4. RetireStake:      Permanently marks the given balance for future withdrawal. Only inactive funds can be retired.
+///                      This action is only possible if:
+///                        (a) if the resulting non retired funds respect the invariant 1 - minimum stake for non retired funds.
```
```suggestion
///                        (a) the resulting non-retired funds respect the invariant 1 - minimum stake for non-retired funds.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -35,17 +38,34 @@ use crate::{
+/// 4. RetireStake:      Permanently marks the given balance for future withdrawal. Only inactive funds can be retired.
+///                      This action is only possible if:
+///                        (a) if the resulting non retired funds respect the invariant 1 - minimum stake for non retired funds.
+///                        (b) the inactive funds are already released (*).
```
```suggestion
///                        (b) the inactive funds are released (*).
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -35,17 +38,34 @@ use crate::{
+///                      total stake or there are no more funds left. On the latter case, this action will result in
+///                      the deletion of the staker.
+///                      This action is only possible if:
+///                         (a) if the resulting total stake respect the invariant 2 - minimum stake for total funds.
```
```suggestion
/// 5. RemoveStake:      Removes retired balance from a staker to outside the staking contract.
///                      Retired balance can always be withdrawn as long as the staker respects the minimum
///                      total stake or all balance is removed. The latter case results in
///                      the deletion of the staker.
///                      This action is only possible if:
///                         (a) the resulting total stake respects the invariant 2 - minimum stake for total funds.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -35,17 +38,34 @@ use crate::{
+///             (invariant 3) Inactivation block height is always associated with some inactive balance:
+///                          `inactive_from` is Some(_) && `inactive_balance` > 0
+///             (invariant 4) When there are no inactive funds, there is no inactivation block height:
+///                          `inactive_from` is None && `inactive_balance` == 0
```
```suggestion
///             (invariant 1) Min stake for non-retired balances:
///                           active + inactive balance must be == 0 or >= minimum stake
///             (invariant 2) Min stake for total staker balance:
///                           active + inactive + retired balance must be == 0 or >= minimum stake
///             (invariant 3) Inactivation block height is always associated with some inactive balance:
///                          `inactive_from` is Some(_) && `inactive_balance` > 0
///             (invariant 4) When there are no inactive funds, there is no inactivation block height:
///                          `inactive_from` is None && `inactive_balance` == 0
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -55,12 +75,17 @@ pub struct Staker {
     /// The stake can only effectively become inactive on the next election block. Thus, this may contain a
     /// future block height.
     pub inactive_from: Option<u32>,
+    /// The staker's retired balance. Retired balance can only be withdrawn, thus this action is irreversible.
```
```suggestion
    /// The staker's retired balance. Retired balance can only be withdrawn, thus retiring is irreversible.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -55,12 +75,17 @@ pub struct Staker {
+    /// The staker's retired balance. Retired balance can only be withdrawn, thus this action is irreversible.
+    /// Only released inactive balance can be retired, so the maximum of the inactive and the validator's jailed
+    /// periods must have passed.
+    /// Once retired, the funds are immediately available to be withdrawn.
```
```suggestion
    /// Once retired, the funds are immediately available to be withdrawn (removed).
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -167,17 +220,31 @@ impl StakingContract {
         // Get the staker.
         let mut staker = store.expect_staker(staker_address)?;
 
+        // Fail if the minimum stake would be violated for the non retired funds (invariant 1).
```
```suggestion
        // Fail if the minimum stake would be violated for the non-retired funds (invariant 1).
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -247,10 +312,10 @@ impl StakingContract {
 
-        // If stake is currently delegated, it must be inactive.
-        if let Some(validator_address) = &staker.delegation {
+        // If stake is currently delegated, than the active balance should be 0.
```
Alternatively, `then`.
```suggestion
        // If stake is currently delegated, the active balance must be 0.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -259,27 +324,14 @@ impl StakingContract {
-                    return Err(AccountError::InvalidForRecipient);
-                }
-            }
+            // Fail if the the funds are in cooldown or jailed.
```
First time the word "cooldown" is used. I suggest using the same words as before to describe the situation. "not yet released" and "locked" come to mind.

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -259,27 +324,14 @@ impl StakingContract {
+            if !self.is_inactive_stake_released(store, &staker, block_number) {
+                debug!(
+                    ?staker_address,
+                    "Tried to update staker with inactive balance locked (cooldown or jail)"
```
```suggestion
                    "Tried to update staker with inactive balance locked or jailed"
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -384,10 +435,10 @@ impl StakingContract {
     /// The active balance will be set immediately.
-    /// The inactive balance will be available to withdraw after a lock up period and, if applicable,
-    /// until the jail period is finished.
+    /// The inactive balance will be available to retire after a lock up period and, if applicable,
```
```suggestion
    /// The inactive balance will be available to retire after a lockup period and, if applicable,
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -384,10 +435,10 @@ impl StakingContract {
-    /// until the jail period is finished.
+    /// The inactive balance will be available to retire after a lock up period and, if applicable,
+    /// after the validator's jail period has finished.
     /// If the staker has already some inactive stake, the corresponding balance will be overwritten and
```
```suggestion
    /// If the staker already has some inactive stake, the corresponding balance will be overwritten and
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -384,10 +435,10 @@ impl StakingContract {
+    /// The inactive balance will be available to retire after a lock up period and, if applicable,
+    /// after the validator's jail period has finished.
     /// If the staker has already some inactive stake, the corresponding balance will be overwritten and
     /// the lock up period will be reset. This corresponds to editing the active stake balance
```
```suggestion
    /// the lockup period will be reset. This corresponds to editing the active stake balance
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -418,6 +469,16 @@ impl StakingContract {
         let old_active_balance = staker.balance;
         let new_inactive_balance = total_balance - new_active_balance;
 
+        // Update the staker's balance.
```
```suggestion
        // Update the staker's balances.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
-        store: &S,
-        staker: &Staker,
+    /// Adds to the retired balance and consequentially changes the inactive balance of the staker.
+    /// The balance can only be retired if the lock up period and associated validator's jail period have passed.
```
```suggestion
    /// The balance can only be retired if the lockup period and associated validator's jail period have passed.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
-        staker: &Staker,
+    /// Adds to the retired balance and consequentially changes the inactive balance of the staker.
+    /// The balance can only be retired if the lock up period and associated validator's jail period have passed.
+    /// The retire fails if the invariant 1 for the non retired stake is violated.
```
```suggestion
    /// The retire fails if the invariant 1 for the non-retired stake is violated.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
+            });
         }
+
+        // Fail if the minimum stake would be violated for the non retired funds (invariant 1).
```
```suggestion
        // Fail if the minimum stake would be violated for non-retired funds (invariant 1).
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
+            return Err(AccountError::InvalidCoinValue);
+        }
+
+        // Fail if the the funds are in cooldown or jailed.
```
"not yet released" or "locked"

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
+        if !self.is_inactive_stake_released(store, &staker, block_number) {
+            debug!(
+                ?staker_address,
+                "Tried to retire stake with inactive balance locked (cooldown or jail)"
```
```suggestion
                "Tried to retire stake with inactive balance locked or jailed"
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
+        let old_inactive_from = staker.inactive_from;
+
+        // Update the staker's balances.
+        staker.inactive_balance -= retire_stake;
```
You created this variable above, might as well use it.
```suggestion
        staker.inactive_balance = new_inactive_balance;
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +551,175 @@ impl StakingContract {
 
-    /// Removes coins from a staker's balance. If the entire staker's balance is removed then the
-    /// staker is deleted.
+    /// Removes coins from the retired balance of the staker.
```
```suggestion
    /// Removes balance from the retired stake of the staker.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +770,106 @@ impl StakingContract {
+    /// This function expects as input the final balances of a given operation.
+    /// Invariants:
+    ///         (1) active + inactive balances should be == 0 or >= Minimum Stake
+    ///         (2) active + inactive + retired balances should be == 0 or >= Minimum Stake
```
```suggestion
    ///         (1) active + inactive balances must be == 0 or >= minimum stake
    ///         (2) active + inactive + retired balances must be == 0 or >= minimum stake
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +770,106 @@ impl StakingContract {
+        inactive_balance: Coin,
+        retired_balance: Coin,
+    ) -> bool {
+        // Check invariant 1: that total non retired balance (active+inactive) does not violate the minimum stake.
```
```suggestion
        // Check invariant 1: total non-retired balances (active+inactive) do not violate the minimum stake.
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +770,106 @@ impl StakingContract {
+            && retired_balance > Coin::ZERO
+            && retired_balance < Coin::from_u64_unchecked(Policy::MINIMUM_STAKE)
+        {
+            log::error!("{:?}", retired_balance);
```
leftover logging from debugging?

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +770,106 @@ impl StakingContract {
+        true
+    }
+
+    /// Returns true if the inactive funds are released both from the cooldown period and any potential jail
```
```suggestion
    /// Returns true if the inactive funds are released both from the lockup period and any potential jail
```

sisou commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +770,106 @@ impl StakingContract {
+        staker: &Staker,
+        block_number: u32,
+    ) -> bool {
+        // If stake is currently delegated we check if funds are released.
+        // We only need to check cooldown of inactive funds if there is a delegation.
+        // This is because the purpose of the cooldown is to ensure that we wait for the reporting window
+        // to finish and thus that the validator will not be jailed for misbehavior.
+        if let Some(validator_address) = &staker.delegation {
+            // Funds are not released if release block height has not passed yet.
+            if let Some(inactive_from) = staker.inactive_from {
```
Didn't you state earlier that an `inactive_from` number is only set when the staker has a delegation? So the check for delegation is unnecessary, as a set `inactive_from` field implies a delegation anyway?

ii-cruz replied:
We can have delegation set to None and have an inactive balance>0 and, thus inactive_from = Some(...).
The invariant is related to inactive_balance and inactive_from, not about the delegation.

ii-cruz replied:
However, if we have inactive funds delegated to nobody, we don't need to bother waiting for lockup or jail periods. 
The jail is self-explanatory, there is no delegation so the delegation cannot be jailed.
For the lockup checking it is unnecessary because the purpose of the lockup is to prevent the staker from retiring stake and potentially removing the stake before the reporting window for the misbehavior of a validator has passed. However, there is no delegation, so nobody relevant can misbehave. Thus waiting for lockup is not needed.

sisou replied:
Ok :+1:

sisou commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -604,13 +640,17 @@ impl AccountTransactionInteraction for StakingContract {
                 let staker = store.expect_staker(&staker_address)?;
 
-                self.can_remove_stake(&store, &staker, block_state.number)?;
+                // Verify that the stake can actually be removed.
+                self.can_remove_stake(&staker, transaction.total_value())?;
+                // Verify that the fee by itself can be removed without violating the minimum stake.
+                self.can_remove_stake(&staker, transaction.fee)?;
```
If the second check is always done, even if the first check passes, then that makes it impossible to remove stake which is exactly 100 NIM with a tx with a fee. Only a 0-fee tx would then be allowed, or you'd have to fund the staker with more NIM beforehand.

ii-cruz replied:
Yes, correct. That was a trade-off we agreed upon to actually enforce the minimum stake despite the possibility of a commit failed tx and its consequential fee deduction.
Since it's an edge case that only happens if a staker is only holding the minimum stake (which is a really small amount) and fees are currently 0 on our network, we decided to enforce the minimum stake at the expense of this edge case.
Otherwise, we would have a minimum stake but it could be broken, similar to the validator deposit. Calling it a minimum stake would be rather pointless, even implementing a "minimum stake" on these conditions would be rather pointless.

sisou replied:
I don't fully understand. Is this code here the ultimate test if the transaction will succeed? Or can it still fail later on?

If it can still fail later on, I agree that this test has to be as it is.

But if this is the ultimate test, and we have already established that the whole tx will go through with the first check (e.g. removing 99 NIM with 1 NIM fee), then why do we still need to test for a failing tx, which is the second check?

ii-cruz replied:
We cannot be 100% sure that the tx will be successful at the reserve balance level. So no, this is not the ultimate test, it can still fail later on. 
It can fail on the receiver for example, by trying to transfer to another contract. That's just one of the cases, there are also some other scenarios where it can still fail later at the commit failed tx despite being accepted here.

sisou replied:
Thanks for clarifying :bulb:

sisou commented on primitives/account/src/account/staking_contract/validator.rs:
```diff
@@ -643,14 +643,35 @@ impl StakingContract {
+        Ok(())
+    }
+
+    /// Checks if a validator is retired and release in order to have a fee or a deposit deduction.
```
```suggestion
    /// Checks if a validator is retired and released in order to have a fee or a deposit deduction.
```

sisou commented on primitives/account/src/account/staking_contract/validator.rs:
```diff
@@ -643,14 +643,35 @@ impl StakingContract {
+
+    /// Checks if a validator is retired and release in order to have a fee or a deposit deduction.
+    /// For this, the validator must be retired and have enough funds and be
+    /// released from jail.
```
This sentence is the same as the previous sentence in meaning.

sisou commented on primitives/account/src/account/staking_contract/validator.rs:
```diff
@@ -643,14 +643,35 @@ impl StakingContract {
         // Check that the validator is retired.
         if !validator.retired {
-            debug!(?validator.address, "Tried to delete active validator");
+            debug!(?validator.address, "Tried to delete a non retired validator");
```
```suggestion
            debug!(?validator.address, "Tried to delete a non-retired validator");
```

sisou commented on primitives/src/policy.rs:
```diff
@@ -105,6 +105,12 @@ impl Policy {
 
+    /// The stake necessary to create a staker in Lunas (1 NIM = 100,000 Lunas).
+    /// This minimum is applied to the amount of:
+    ///     - non retired stake
```
```suggestion
    ///     - non-retired stake
```

sisou commented on primitives/transaction/src/account/staking_contract/structs.rs:
```diff
@@ -161,17 +165,23 @@ impl IncomingStakingTransactionData {
-                // No checks needed.
+                // Adding stake should be at least greater than 0.
+                if transaction.value.is_zero() {
+                    warn!("Add stake transactions must actually have higher than 0 value. The offending transaction is the following:\n{:?}", transaction);
```
```suggestion
                    warn!("Add stake transactions must have higher than 0 value. The offending transaction is the following:\n{:?}", transaction);
```

sisou commented on rpc-client/src/subcommands/transactions_subcommands.rs:
```diff
@@ -72,8 +72,9 @@ pub enum TransactionCommand {
-    /// Sends a staking transaction from the address of a given `wallet` to a given `staker_address`.
-    Stake {
+    /// Sends a `add_stake` transaction from the address of a given `wallet` to a given `staker_address`.
+    /// This transaction result must have the sum of active and inactive stake is >= minimum stake, otherwise it fails.
```
```suggestion
    /// This transaction must result in the sum of active and inactive stake to be >= minimum stake, otherwise it fails.
```

sisou commented on rpc-client/src/subcommands/transactions_subcommands.rs:
```diff
@@ -123,23 +124,46 @@ pub enum TransactionCommand {
+    /// This transaction can only result in all funds being retired or the sum of active and new inactive
+    /// balances being >= minimum stake. Otherwise it fails.
+    RetireStake {
+        /// The fee will be payed by this wallet if any is provided. In such case the sender wallet must be unlocked prior to this action.
```
```suggestion
        /// The fee will be paid by this wallet if any is provided. In such case the sender wallet must be unlocked prior to this action.
```

sisou commented on rpc-client/src/subcommands/transactions_subcommands.rs:
```diff
@@ -123,23 +124,46 @@ pub enum TransactionCommand {
+    /// balances being >= minimum stake. Otherwise it fails.
+    RetireStake {
+        /// The fee will be payed by this wallet if any is provided. In such case the sender wallet must be unlocked prior to this action.
+        /// If absent the fee is payed by the stakers account.
```
```suggestion
        /// If absent, the fee is paid by the staker's account.
```

sisou commented on test-utils/src/transactions.rs:
```diff
@@ -219,10 +222,25 @@ impl<R: Rng + CryptoRng> TransactionsGenerator<R> {
+        } else if matches!(sender, OutgoingType::RemoveStake) {
+            value = Coin::from_u64_unchecked(2 * Policy::MINIMUM_STAKE)
+                .checked_sub(fee)
+                .expect("Fee should not be higher than staker stake");
```
```suggestion
                .expect("Fee must not be higher than staker stake");
```

sisou commented on transaction-builder/src/lib.rs:
```diff
@@ -883,7 +883,7 @@ impl TransactionBuilder {
     ///  - `delegation`:            The (optional) delegation to a validator.
     ///  - `value`:                 The value for the initial stake. This is sent from the account
-    ///                             belonging to `key_pair`.
+    ///                             belonging to `key_pair`. Value must be >= to minimum stake.
```
```suggestion
    ///                             belonging to `key_pair`. Value must be >= minimum stake.
```

sisou commented on transaction-builder/src/lib.rs:
```diff
@@ -925,7 +925,7 @@ impl TransactionBuilder {
     }
 
-    /// Creates a staking transaction from the address of a given `key_pair` to a specified
+    /// Creates a transaction to add more stake from the address of a given `key_pair` to a specified
```
```suggestion
    /// Creates a transaction to add stake from the address of a given `key_pair` to a specified
```

sisou commented on transaction-builder/src/lib.rs:
```diff
@@ -934,7 +934,8 @@ impl TransactionBuilder {
     ///  - `value`:                 The value of the stake. This is sent from the account
-    ///                             belonging to `key_pair`.
+    ///                             belonging to `key_pair`. This transaction can fail if the value violates
+    ///                             minimum non retired stake
```
```suggestion
    ///                             belonging to `key_pair`. This transaction can fail if the resulting staker violates
    ///                             the minimum non-retired stake.
```

sisou commented on transaction-builder/src/recipient/staking_contract.rs:
```diff
@@ -160,6 +160,18 @@ impl StakingRecipientBuilder {
+    /// This method allows to retire the staker's balance. The retired funds are deducted from the inactive
+    /// stake. Retiring funds is only allowed after the inactive funds are released (i.e. cooldown and jail
+    /// periods have passed). Additionally, minimum total stake for non retired funds must still be respected.
+    /// It needs to be signed by the key pair corresponding to the staker address.
```
```suggestion
    /// This method allows to retire stake from the staker's inactive balance.
    /// Retiring funds is only allowed after the inactive funds are released (i.e. lock-up and jail
    /// periods have passed). Additionally, the minimum total stake for non-retired funds must still be respected.
    /// It needs to be signed by the key pair corresponding to the staker address.
```

sisou commented on web-client/example/module.js:
```diff
@@ -148,14 +148,14 @@ init().then(async () => {
      * @returns {Promise<string>}
      */
-    window.sendUnstakeTransaction = async (privateKey, amount, fee = 0) => {
+    window.sendRemoveStakeTransaction = async (privateKey, amount, fee = 0) => {
```
For this to make sense, we also need `window.sendSetActiveStakeTransaction` and  `window.sendRetireStakeTransaction` methods now.

ii-cruz replied:
conclusion: this needs updating but later. this is an example file and is missing other transactions besides the ones already mentioned.

sisou commented on web-client/src/primitives/transaction_builder.rs:
```diff
@@ -259,6 +256,40 @@ impl TransactionBuilder {
         Ok(Transaction::from_native(tx))
     }
 
+    /// Remove stake from the staking contract and transfers `value` amount of luna (NIM's smallest unit)
```
```suggestion
    /// Removes stake from the staking contract and transfers `value` amount of luna (NIM's smallest unit)
```

sisou commented on web-client/src/primitives/transaction_builder.rs:
```diff
@@ -228,22 +225,22 @@ impl TransactionBuilder {
-    pub fn new_set_active_stake(
+    /// Throws when the numbers given for fee and `retire_stake` do not fit within a u64 or the networkId is unknown.
+    #[wasm_bindgen(js_name = newRetireStake)]
+    pub fn retire_stake(
```
```suggestion
    pub fn new_retire_stake(
```

sisou reviewed (approved):
The logic seems to mostly be correct :+1:
I have however several wording and documentation clarity suggestions.

hrxi commented on primitives/account/tests/accounts.rs:
```diff
@@ -988,6 +988,16 @@ fn can_revert_transactions() {
 
                     let receipts = accounts.test(&[tx], &[], &block_state);
                     if fail_sender || fail_recipient {
+                        for aa in receipts.transactions.iter() {
```
```suggestion
                        for receipt in receipts.transactions.iter() {
```

ii-cruz replied:
Ups those were debug prints, took them out :pray:

hrxi commented on primitives/account/tests/accounts.rs:
```diff
@@ -988,6 +988,16 @@ fn can_revert_transactions() {
+                                    error!("Ok {:?}", tx);
+                                }
+                                OperationReceipt::Err(tx, reason) => {
+                                    error!("Ok {:?} - {:?}", tx, reason);
```
```suggestion
                                    error!("Err {:?} - {:?}", tx, reason);
```

hrxi commented on primitives/account/tests/staking_contract/staker.rs:
```diff
@@ -1602,6 +1606,44 @@ fn unstake_works() {
 }
 
+#[test]
+fn unstake_does_not_allow_nim_stake_to_be_violated() {
```
`nim_stake` ^^ (Gone in a later commit)

hrxi commented on primitives/transaction/src/account/staking_contract/structs.rs:
```diff
@@ -161,17 +165,23 @@ impl IncomingStakingTransactionData {
-                // No checks needed.
+                // Adding stake should be at least greater than 0.
+                if transaction.value.is_zero() {
+                    warn!("Add stake transactions must have higher than 0 value. The offending transaction is the following:\n{:?}", transaction);
```
```suggestion
                    warn!("Add stake transactions must have positive value. The offending transaction is the following:\n{:?}", transaction);
```
or non-zero.

hrxi commented on primitives/transaction/src/account/staking_contract/structs.rs:
```diff
@@ -181,38 +191,38 @@ impl IncomingStakingTransactionData {
+            } => {
+                // Check that retire is greater than 0.
+                if retire_stake.is_zero() {
+                    warn!("Retire stake transactions must have higher than 0 amount to retire. The offending transaction is the following:\n{:?}", transaction);
```
```suggestion
                    warn!("Retire stake transactions must retire a non-zero amount of stake. The offending transaction is the following:\n{:?}", transaction);
```

hrxi commented on primitives/account/tests/staking_contract/staker.rs:
```diff
@@ -346,388 +425,505 @@ fn stake_works() {
     let mut db_txn = staker_setup.env.write_transaction();
     let mut db_txn = (&mut db_txn).into();
-    let mut data_store_write = data_store.write(&mut db_txn);
+    _ = data_store.write(&mut db_txn);
```
What does this do? (Also the further occurrences in this file.)

ii-cruz replied:
This is a bit annoying. So without that write, the .into() from the line above infers the wrong type because the first operation performed in these tests using the `db_txn` is a `get_staker`. This is a read operation on the datastore and thus when we use it to write (on the commit incoming transaction) we have a type mismatch. 
This is totally a hardcoded way of dealing with the problem, let me know if there is a cleaner way.

hrxi replied:
I guess a type annotation on the second `db_txn` would work?

```rs
use nimiq_trie::WriteTransactionProxy;

let mut db_txn = staker_setup.env.write_transaction();
let mut db_txn: WriteTransactionProxy = (&mut db_txn).into();
```

ii-cruz replied:
This was so damn stupid xD I had tried it but the rust analyzer was complaining still, I probably had the wrong import. 🙈

hrxi commented on test-utils/src/transactions.rs:
```diff
@@ -215,14 +218,29 @@ impl<R: Rng + CryptoRng> TransactionsGenerator<R> {
+        } else if matches!(recipient, IncomingType::SetActiveStake) {
+            value = Coin::ZERO;
+        } else if matches!(recipient, IncomingType::RetireStake) {
+            value = Coin::from_u64_unchecked(Policy::MINIMUM_STAKE);
```
This looks like it could benefit from an actual `match` instead of many `if matches!`.

hrxi reviewed (approved):
Looks generally good to me. I added a couple of nits.

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -78,6 +101,30 @@ impl StakingContract {
+    }
+
+    /// Creates a new staker with retired funds.
+    fn create_staker_with_retired(
```
```suggestion
    fn create_staker_with_retired_balance(
```

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +768,105 @@ impl StakingContract {
+    /// Invariants:
+    ///         (1) active + inactive balances must be == 0 or >= minimum stake
+    ///         (2) active + inactive + retired balances must be == 0 or >= minimum stake
+    fn is_min_stake_enforced(
```
```suggestion
    fn is_min_stake_satisfied(
```

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -637,6 +768,105 @@ impl StakingContract {
+        retired_balance: Coin,
+    ) -> bool {
+        // Check invariant 1: that total non-retired balances (active+inactive) do not violate the minimum stake.
+        let non_retired_balances = active_balance + inactive_balance;
```
```suggestion
        let non_retired_balance = active_balance + inactive_balance;
```

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +549,175 @@ impl StakingContract {
+        if !self.is_inactive_stake_released(store, &staker, block_number) {
+            debug!(
+                ?staker_address,
+                "Tried to retire stake with inactive balance locked or jail"
```
```suggestion
                "Tried to retire stake with inactive balance locked or jailed"
```

styppo commented on test-utils/src/transactions.rs:
```diff
@@ -212,25 +215,51 @@ impl<R: Rng + CryptoRng> TransactionsGenerator<R> {
         // Ensure that the value is the validator deposit if we are creating a validator.
-        if matches!(recipient, IncomingType::CreateValidator) {
-            value = Coin::from_u64_unchecked(Policy::VALIDATOR_DEPOSIT);
+        // Ensure that the value is the minimum stake is respected in the staker operations.
```
```suggestion
        // Ensure that the minimum stake is respected in staker operations.
```

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -35,17 +37,31 @@ use crate::{
+///                        (b) the inactive funds are released (*).
+/// 5. RemoveStake:      Removes the retired balance from a staker to outside the staking contract.
+///                      The retired balance can always be withdrawn as long as it removes all retired stake.
+///                      In case of total sake = 0 results in the deletion of the staker.
```
```suggestion
///                      If the staker's total balance drops to 0, the staker is deleted.
```

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -35,17 +37,31 @@ use crate::{
 ///
-/// Create, Stake, SetActiveStake, and Update are incoming transactions to the staking contract.
-/// Unstake is an outgoing transaction from the staking contract.
+/// Create, Stake, SetActiveStake, Update and RetireStake are incoming transactions to the staking contract.
```
```suggestion
/// Create, AddStake, SetActiveStake, Update and RetireStake are incoming transactions to the staking contract.
```

styppo commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -459,35 +479,62 @@ impl AccountTransactionInteraction for StakingContract {
+                    staker.retired_balance = new_retired_balance;
+                    self.balance -= transaction.fee;
+
+                    // Update the validator entry.
```
```suggestion
                    // Update the staker entry.
```

styppo commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -459,35 +479,62 @@ impl AccountTransactionInteraction for StakingContract {
                     )
-                    .map(|receipt| Some(receipt.into()))?;
+                } else {
+                    // Update the balances of the validator and staking contract.
```
```suggestion
                    // Update the balances of the staker and staking contract.
```

styppo commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -459,35 +479,62 @@ impl AccountTransactionInteraction for StakingContract {
+                    new_retired_balance,
+                )?;
+
+                let receipt = if new_retired_balance.is_zero() {
```
This must check if the **total** balance is zero.

What was the reason to change this from the previous implementation that simply called through to `remove_stake` in all cases and did not need this `if` at all? Generating the `DeleteStaker` event by checking if the staker has been removed from the store seems a bit simpler and with less duplication.

styppo commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -459,35 +479,62 @@ impl AccountTransactionInteraction for StakingContract {
+                    None
+                };
 
                 tx_logger.push_log(Log::StakerFeeDeduction {
```
I think it would be more consistent to log `StakerFeeDeduction` *before* `DeleteStaker`. The fee is deducted before the staker is deleted.

styppo commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -553,21 +596,47 @@ impl AccountTransactionInteraction for StakingContract {
+                self.balance += transaction.fee;
+
+                // Update the validator entry.
+                store.put_staker(&staker_address, staker);
```
See comment above regarding simply calling through to `revert_remove_stake` here as it was before.

styppo commented on primitives/account/src/account/staking_contract/traits.rs:
```diff
@@ -553,21 +596,47 @@ impl AccountTransactionInteraction for StakingContract {
+                // Update the validator entry.
+                store.put_staker(&staker_address, staker);
+
+                if store.get_staker(&staker_address).is_none() {}
```
Unnecessary line

styppo commented on primitives/account/src/account/staking_contract/staker.rs:
```diff
@@ -500,96 +538,162 @@ impl StakingContract {
-
-        // Update our balance.
+        // Update balances.
+        staker.retired_balance = Coin::ZERO;
```
It would be more consistent to subtract instead of setting to zero.

ii-cruz committed:
Min stake for non retired and total stake, despite fee deductions.

Enforcing minimum total stake despite fee deduction.
Adding Minimum active + inactive stake invariant.
Remove stake must remove total retired stake.
Fixing staking contract signalling transactions confusion between value and balances.
Renames stake to add_stake
General staking contract and failed transactions cleanup.
Fixes the RPC commands and comments.
Fixes the comments.
Adding tests.
